sotodlib/flags.py

```diff
@@ -14,6 +15,222 @@
+    """
+    
+    # Get daz and time 
+    if az == None : 
+        az = np.copy(aman.boresight.az)
+    daz = np.diff(az)
+    daz = np.append(daz,daz[-1])
+    time = aman.timestamps - aman.timestamps[0]
+
+    # derive approx daz
+    approx_daz = np.mean(daz[daz>np.percentile(daz, 95)])
+
+    # derive approx aprox daz/dt and rough number of samples in one_scan period 
+    dt = np.mean(np.diff(aman.timestamps))
+    approx_dazdt = approx_daz / dt
+    approx_samps_onescan = int(np.ptp(az) / approx_dazdt / dt)
+
+    # Make a step-function like matched filter
+    kernel_size = 400
+    kernel = np.zeros(kernel_size)
+    # normalize with approx_daz so that peaks height be ~1
+    kernel[:kernel_size//2] = 1/approx_daz
+    kernel[kernel_size//2:] = - 1/approx_daz
+    kernel /= kernel_size
+
+    # convolve signal with the kernel
+    matched = np.convolve(np.diff(aman.boresight.az), kernel, mode='same')
+    matched[:kernel_size] = 0
+    matched[-kernel_size:] = 0 
+
+    # find peaks in matched daz, height=0.1, distance=approx_samps_onescan//3 would be robust enough
+    peaks, properties = find_peaks(matched, height=0.1, distance=approx_samps_onescan//3)
+
+    # turn around flag 
+    turnaround_mask = np.zeros_like(aman.timestamps,dtype=bool)
+
+    ### flagging
+    
+    mean_length = np.mean(np.diff(peaks))
+    prev = 0    
+    nbuffer = int(t_buffer/dt) # A finite buffer is needed to avoid the fluctuation of daz/dt
+
+    # Check the beggining part 
+    first_is_positive = False
+    if peaks[0] < mean_length//2 : 
+        first_is_positive = True
+
+    # search for edges of square waves looking at the peaks
+    for p in peaks : 
+        start = prev
+        end = p
+        length = p-prev
+        middle = start + length//2
+        prev = p
+
+        turnaround_mask[middle-nbuffer:middle+nbuffer] = True
+        turnaround_mask[prev-nbuffer:prev+nbuffer] = True
+        
+    # we don't know the end of the last square wave, so use the mean of wave width
+    start = prev
+    end = prev + mean_length
+    middle = start + mean_length//2
+    
+    turnaround_mask[int(middle-nbuffer):int(middle+nbuffer)] = True
+    turnaround_mask[int(prev-nbuffer):int(prev+nbuffer)] = True    
+    
+    # treatment of the begining
+    if first_is_positive : 
+        turnaround_mask[:peaks[0]+nbuffer] = True
+
+    else : 
+        turnaround_mask[:int(peaks[0]-mean_length//2)] = True
+        
+    ta_flag = Ranges.from_bitmask(turnaround_mask)
+
+    if merge:
+        if name in aman.flags and not overwrite:
+            raise ValueError("Flag name {} already exists in aman.flags".format(name))
+        elif name in aman.flags:
+            aman.flags[name] = ta_flag
+        else:
+            aman.flags.wrap(name, ta_flag)
+    return ta_flag
+    
+
+def get_leftright_flags_by_dazdt(
+    aman, t_buffer=1.8, w_turnaround=True, az=None, merge=True, overwrite=False, name=["left", "right"]
```

Comment:
Can we integrate `get_leftright_flags` and `get_turnaround_flags` since some of lines are duplicated.